File: python/euler/bin/p0049.py
# project euler: problem 49

from functools import reduce

from euler.lib.prime import primes


def make_prime_tbl(ndigits: int) -> dict[str, list[int]]:
    p_tbl: dict[str, list[int]] = {}

    for p in primes(10 ** (ndigits - 1), 10**ndigits):
        key = ''.join(sorted(str(p)))
        v = p_tbl.get(key, [])
        v.append(p)
        p_tbl[key] = v

    return p_tbl


def compute(ndigits: int) -> str:
    p_tbl = make_prime_tbl(ndigits)
    for lst in p_tbl.values():
        if (length := len(lst)) < 3:
            continue
        for i in range(length - 2):
            for j in range(i + 1, length - 1):
                tmp = lst[j] * 2 - lst[i]  # (x + a) * 2 - x = x + 2a
                if tmp in lst and lst[i] != 1487 and lst[j] != 4817:
                    return reduce(lambda x, y: x + y, map(str, [lst[i], lst[j], tmp]))

    raise RuntimeError('unreachable!')


def solve() -> str:
    return compute(4)

File: typescript/euler/bin/p0049.ts
// project euler: problem 49

import { combinations } from "combinatorics/mod.ts";
import { isPrime, primes } from "../lib/prime.ts";

export const compute = (limit: number): string => {
  const isPerm = (p1: number, p2: number, p3: number): boolean => {
    const s1 = String(p1).split("").sort().join("");
    const s2 = String(p2).split("").sort().join("");
    const s3 = String(p3).split("").sort().join("");

    return s1 === s2 && s1 === s3;
  };

  const primeLst = primes(1000, limit);
  for (const [i, j] of combinations(primeLst, 2)) {
    const m = Math.trunc((i + j) / 2);
    if (isPrime(m) === true && isPerm(i, m, j) === true) {
      if (i + j === 1487 + 8147) {
        continue;
      }
      return String(i) + String(m) + String(j);
    }
  }

  throw new Error("not reached");
};

export const solve = (): string => compute(9_999);
